Add examples for doctest to optuna/trial.py

nuka137 opened the pull request:
This PR add doctest example to optuna/trial.py. (Related Issue: #734)

Some simple assertions are added to each example.
Classifiers in examples are left as original.
Each classifier uses simple input array generated by `numpy.random` .

**Note**

Because of numpy issue, below warning is raised when run `make doctest` to optuna/trial.py.
Should this warning message be suppressed?

```
/opt/conda/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 216, got 192
  return f(*args, **kwds)
```

toshihikoyanase commented on optuna/trial.py:
```diff
@@ -357,21 +411,32 @@ def report(self, value, step=None):
         Example:
 
-            Report intermediate scores of `SGDClassifier <https://scikit-learn.org/stable/modules/
-            generated/sklearn.linear_model.SGDClassifier.html>`_ training
```
IMO, this description is useful for users because it clarifies the usage of the `report` function. What do you think of it?

nuka137 replied:
Thanks!
Yes, I think this document is useful for users.
But I'm not sure why #702 deleted the document on Example.
Is there any criteria to delete documents?

toshihikoyanase replied:
I guess that is because the description in the previous commit is not valid. 

>  Get an objective value and a value of parameter ``x`` in the first row.

In the previous example, the code actually obtains the objective value and parameter value as follows:

```python
>>> df.value[0]	
        0.0
>>> df.params.x[0]
        1.0
```

But they are replaced by the following assertions in the new example:

```python
assert isinstance(df, pandas.DataFrame)
assert df.shape[0] == 3  # n_trials.
```

> Is there any criteria to delete documents?

I don't think we have such criteria because we just started using the doctest.

@hvy @sile @g-votte 

Do you have any comments about the description (or title?) of the doctest code?

toshihikoyanase commented on optuna/trial.py:
```diff
@@ -357,21 +411,32 @@ def report(self, value, step=None):
+                import optuna
+                import numpy as np
+                from sklearn.linear_model import SGDClassifier
+                from sklearn.model_selection import train_test_split
```
This example has more than 20 lines, and `trial.report` may not be emphasized enough. 
It is controversial, but I think we can use the `testsetup` directive to hide unimportant lines such as the module imports and the dataset preparation.


Here's an example code and rendering result.

```python
            .. testsetup::

                import optuna
                import numpy as np
                from sklearn.linear_model import SGDClassifier
                from sklearn.model_selection import train_test_split

                np.random.seed(seed=0)
                X = np.random.randn(50).reshape(-1, 1)
                y = np.random.randint(0, 2, 50)
                X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

            .. testcode::

                def objective(trial):
```

![image](https://user-images.githubusercontent.com/3255979/73531665-485eac00-445e-11ea-9540-01a1569e9fd0.png)

nuka137 replied:
I think your solution is better than mine. Output document is more understandable.
BTW, should I apply this to other examples too?

toshihikoyanase replied:
Thanks!

> BTW, should I apply this to other examples too?

I don't think so. The `trial.report` example is essentially long because it requires a dataset and a machine learning model to be viable and realistic. On the other hand, the suggest APIs such as `trial.suggest_uniform` and `trial.suggest_int` do not require such preparation. So, I think all examples for them can be simplified like the `trial.suggest_uniform` example and they will be short enough.

nuka137 commented:
Thanks for your reply.
I reverted the all example document, and moved some codes to `testsetup` from `testcode` about `trial.report` test.
And I also noticed that some original example document uses colon `:` instead of period `.` .
So I fixed them to be consistent.

toshihikoyanase commented on optuna/trial.py:
```diff
@@ -197,13 +204,28 @@ def suggest_loguniform(self, name, low, high):
-                >>>     ...
+                def objective(trial):
+                    c = trial.suggest_loguniform('c', 1e-5, 1e2)
+                    assert 1e-5 <= c < 1e2
```
Please remove the assertion to simplify the example. It looks like a test code rather than an example because the range is inputted right before this line.

```suggestion
```

toshihikoyanase commented on optuna/trial.py:
```diff
@@ -243,13 +265,28 @@ def suggest_discrete_uniform(self, name, low, high, q):
-                >>>     ...
+                def objective(trial):
+                    subsample = trial.suggest_discrete_uniform('subsample', 0.1, 1.0, 0.1)
+                    assert 0.1 <= subsample <= 1.0
```
ditto.
```suggestion
```

toshihikoyanase commented on optuna/trial.py:
```diff
@@ -283,13 +320,29 @@ def suggest_int(self, name, low, high):
+
+                def objective(trial):
+                    n_estimators = trial.suggest_int('n_estimators', 50, 400)
+                    assert 50 <= n_estimators <= 400
```
ditto.
```suggestion
```

toshihikoyanase commented on optuna/trial.py:
```diff
@@ -320,13 +373,29 @@ def suggest_categorical(self, name, choices):
+
+                def objective(trial):
+                    kernel = trial.suggest_categorical('kernel', ['linear', 'poly', 'rbf'])
+                    assert kernel in ['linear', 'poly', 'rbf']
```
ditto.
```suggestion
```

toshihikoyanase commented on optuna/trial.py:
```diff
@@ -197,13 +204,28 @@ def suggest_loguniform(self, name, low, high):
+                    np.random.seed(seed=0)
+                    X = np.random.randn(50).reshape(-1, 1)
+                    y = np.random.randint(0, 2, 50)
+                    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
```
Sorry to change my mind, but this dataset preparation is commonly seen in other examples, so please hide it into the `.. testsetup::` block?

Current:
![image](https://user-images.githubusercontent.com/3255979/74005348-e05b1900-49bb-11ea-9fd1-d18228e03218.png)


Proposal:
![image](https://user-images.githubusercontent.com/3255979/74005267-a558e580-49bb-11ea-866a-1c33ed79fc2e.png)

toshihikoyanase commented on optuna/trial.py:
```diff
@@ -283,13 +320,29 @@ def suggest_int(self, name, low, high):
+                    np.random.seed(seed=0)
+                    X = np.random.randn(50).reshape(-1, 1)
+                    y = np.random.randint(0, 2, 50)
+                    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
```
Please hide this dataset preparation, as I suggested in the `trial.suggest_loguniform` example.

toshihikoyanase commented on optuna/trial.py:
```diff
@@ -320,13 +373,29 @@ def suggest_categorical(self, name, choices):
+                    np.random.seed(seed=0)
+                    X = np.random.randn(50).reshape(-1, 1)
+                    y = np.random.randint(0, 2, 50)
+                    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
```
Please hide this dataset preparation, as I suggested in the `trial.suggest_loguniform` example.

toshihikoyanase commented on optuna/trial.py:
```diff
@@ -437,16 +523,22 @@ def set_user_attr(self, key, value):
         Example:
 
-            Save fixed hyperparameters of neural network training:
+            Save fixed hyperparameters of neural network training.
```
This title is not consistent with the code. Is it possible to actually train the neural networks? Or I think we can work on it in the follow-up PR.

nuka137 commented:
@toshihikoyanase 

Thanks for your review!
I fixed issues except for Neural Network Example.
Before trying to implement NN Example, I would like to check the direction of implementation.
scikit-learn has `NLPClassifier` but there is no `dropout_rate` parameter in it.
Should I use another machine framework library such as PyTorch, TensorFlow? or choose another parameter instead of `dropout_rate` parameter?

And I noticed setup codes are still exist in `suggest_discrete_uniform` which you did not mention.
If it is not your intention, I will fix this too. Any suggestions?

toshihikoyanase commented:
@nuka137 
Thank you for updating the examples.

> Before trying to implement NN Example, I would like to check the direction of implementation.
> scikit-learn has NLPClassifier but there is no dropout_rate parameter in it.
> Should I use another machine framework library such as PyTorch, TensorFlow? or choose another parameter instead of dropout_rate parameter?

IMO, `scikit-learn` is preferable. If we use PyTorch or TensorFlow, much effort to keep the examples up-to-date because they are actively developed and their updates sometimes contain compatibility-breaking changes.

I think that we can replace `dropout` with other hyperparameters. How about tuning te [`momentum` of `MLPClassifier`](https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html)? `momentum` is also tuned in [examples/mlflow/keras_mlflow.py](https://github.com/optuna/optuna/blob/master/examples/mlflow/keras_mlflow.py).

> And I noticed setup codes are still exist in suggest_discrete_uniform which you did not mention.

I'm sorry, but I overlooked it. Please update it as well.

nuka137 commented:
@toshihikoyanase 

Thanks for your instruction!
I added `MLPClassifier` examples to `suggest_uniform` and `set_user_attr`.
Also, I moved the setup codes to `testsetup` in `suggest_discrete_uniform`.

toshihikoyanase commented on optuna/trial.py:
```diff
@@ -156,14 +156,33 @@ def suggest_uniform(self, name, low, high):
+                def objective(trial):
+                    momentum = trial.suggest_uniform('momentum', 0.0, 1.0)
+                    clf = MLPClassifier(hidden_layer_sizes=(100, 50), momentum=momentum,
+                                        tol=1e-3, random_state=0)
```
Please add `solver='sgd'` to arguments of `MLPClassifier()` because `momentum` is not used by the default solver (adam). c.f., https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html#sklearn.neural_network.MLPClassifier

In addition, can we remove the `tol` argument to simplify the example?

nuka137 replied:
Thanks! I fixed them.

nuka137 commented:
@toshihikoyanase 

Thanks for your review!
I fixed the issues you mentioned.
Could you check them again?

hvy commented:
Let me slightly update the title of this PR to match the format of our release notes.
